src/codewhisperer/service/featureConfigProvider.ts

```diff
@@ -0,0 +1,87 @@
+    [testFeatureName, new FeatureContext(testFeatureName, 'CONTROL', { stringValue: 'testValue' })],
+])
+
+export class FeatureConfigProvider {
+    private featureConfigs = new Map<string, FeatureContext>()
+
+    static #instance: FeatureConfigProvider
+
+    constructor() {
+        this.fetchFeatureConfigs()
+
+        setInterval(this.fetchFeatureConfigs.bind(this), featureConfigPollIntervalInMs)
+    }
+
+    public static get instance() {
+        return (this.#instance ??= new this())
+    }
+
+    async fetchFeatureConfigs(): Promise<void> {
+        if (AuthUtil.instance.isConnectionExpired()) {
+            return
+        }
+
+        getLogger().debug('Fetching feature configs')
+        try {
+            const response = await client.listFeatureEvaluations()
+
+            // Overwrite feature configs from server response
+            response.featureEvaluations.forEach(evaluation => {
+                this.featureConfigs.set(
+                    evaluation.feature,
+                    new FeatureContext(evaluation.feature, evaluation.variation, evaluation.value)
+                )
+            })
+        } catch (e) {
+            getLogger().debug('Error when fetching feature configs', e)
+        }
+        getLogger().debug(`Current feature configs: ${this.getFeatureConfigsTelemetry()}`)
+    }
+
+    getFeatureConfigsTelemetry(): string {
+        return `{${Array.from(this.featureConfigs.entries())
+            .map(([name, context]) => `${name}: ${context.variation}`)
+            .join(', ')}}`
+    }
+
+    // TODO: for all feature variations, define a contract that can be enforced upon the implementation of
+    // the business logic.
+    // When we align on a new feature config, client-side will implement specific business logic to utilize
+    // these values by:
+    // 1) Add an entry in featureDefinitions, which is <feature_name> to <feature_context>.
+    // 2) Add a function with name `getXXX`, where XXX refers to the feature name.
+    // 3) Specify the return type: One of the return type string/boolean/Long/Double should be used here.
+    // 4) Specify the key for the `getFeatureValueForKey` helper function which is the feature name.
+    // 5) Specify the corresponding type value getter for the `FeatureValue` class. For example,
+    // if the return type is Long, then the corresponding type value getter is `longValue()`.
+    // 6) Add a test case for this feature.
+    // 7) In case `getXXX()` returns undefined, it should be treated as a default/control group.
+    getTestFeature(): string | undefined {
+        return this.getFeatureValueForKey(testFeatureName).stringValue
+    }
+
+    // Get the feature value for the given key.
+    // In case of a misconfiguration, it will return a default feature value of Boolean true.
+    private getFeatureValueForKey(name: string): FeatureValue {
+        return this.featureConfigs.get(name)?.value ?? featureDefinitions.get(name)?.value ?? { boolValue: true }
```

Comment:
seems like 
```
type FeatureValue: {
   stringValue?: string
   boolValue?: boolean
   doulbeValue?: boolean
   .....
}
```
(I just check the sdk defnition https://github.com/aws/aws-toolkit-vscode/pull/3989/files#diff-2bd14827c41c712bcaef58803209651343cc08861887376fe80f1285e1b1c5d9R328,  not sure about the exact interface)

a qq, what if the other value is read, for example `stringValue`? so here we just let it to be `undefined` correct?

Reply:
It will be undefined.
eventually, the `getXXX()`  for each feature will always return a nullable value, which the caller is responsible for handling.